File: content/09/code/js/06-use-person-parameter-type-everywhere/features/step_definitions/steps.js
const { Given, When, Then, Before } = require("@cucumber/cucumber")
const { assertThat, contains, is, not, containsInAnyOrder } = require("hamjest")
const assert = require("assert")

const { Network } = require("../../src/shouty")

const default_range = 100

Before(function () {
  this.messagesShoutedBy = {}
  this.network = new Network(default_range)
})

Given("the range is {int}", function (range) {
  this.network = new Network(range)
})

Given('{person} is located at {int}', function (person, location) {
  person.moveTo(location)
})

Given("{person} has bought {int} credits", function (sean, credits) {
  sean.credits = credits
})

When("{person} shouts", function (shouter) {
  shouter.shout("Hello, world")
})

When(
  "{person} shouts {int} messages containing the word {string}",
  function (shouter, count, word) {
    for (let i = 0; i < count; i++) {
      const message = `A message containing the word ${word}`
      shouter.shout(message)
      if (!this.messagesShoutedBy[shouter.name]) this.messagesShoutedBy[shouter.name] = []
      this.messagesShoutedBy[shouter.name].push(message)
    }
  }
)

When("{person} shouts {int} over-long messages", function (shouter, count) {
  for (let i = 0; i < count; i++) {
    const baseMessage = `A message from ${shouter.name} that is 181 characters long `
    const message = baseMessage + "x".repeat(181 - baseMessage.length)
    shouter.shout(message)
    if (!this.messagesShoutedBy[shouter.name]) this.messagesShoutedBy[shouter.name] = []
    this.messagesShoutedBy[shouter.name].push(message)
  }
})

When("{person} shouts {string}", function (shouter, message) {
  shouter.shout(message)
  if (!this.messagesShoutedBy[shouter.name]) this.messagesShoutedBy[shouter.name] = []
  this.messagesShoutedBy[shouter.name].push(message)
})

When("{person} shouts the following message", function (shouter, message) {
  shouter.shout(message)
  if (!this.messagesShoutedBy[shouter.name]) this.messagesShoutedBy[shouter.name] = []
  this.messagesShoutedBy[shouter.name].push(message)
})

Then("{person} should hear a shout", function (listener) {
  assertThat(listener.messagesHeard().length, is(1))
})

Then("{person} should not hear a shout", function (listener) {
  assertThat(listener.messagesHeard().length, is(0))
})

Then("{person} hears the following messages:", function (listener, expectedMessages) {
  let actualMessages = listener
    .messagesHeard()
    .map((message) => [message])

  assert.deepEqual(actualMessages, expectedMessages.raw())
})

Then("{person} hears all {person}'s messages", function (listener, shouter) {
  assert.deepEqual(
    listener.messagesHeard(),
    this.messagesShoutedBy[shouter.name]
  )
})

Then("{person} should have {int} credits", function (person, expectedCredits) {
  assertThat(person.credits, is(expectedCredits))
})

File: content/bdd-with-cucumber/09/code/ruby/07-add-a-way-to-change-the-network-s-range/features/step_definitions/steps.rb
DEFAULT_RANGE = 100

Given "the range is {int}" do |range|
  self.network = Shouty::Network.new(range)
end

Given "{person} is located at {int}" do |person, location|
  person.move_to(location)
end

Given('{person} has bought {int} credits') do |person, credits|
  person.credits = credits
end

When "{person} shouts" do |shouter|
  shout from: shouter, message: "Hello, world"
end

When('{person} shouts {string}') do |shouter, message|
  shout from: shouter, message: message
end

When "{person} shouts the following message" do |shouter, message|
  shout from: shouter, message: message
end

When "{person} shouts {int} over-long messages" do |shouter, count|
  count.times do
    base_message = "A message from #{shouter.name} that is 181 characters long "
    message = base_message + "x" * (181 - base_message.size)
    shout from: shouter, message: message
  end
end

When '{person} shouts {int} messages containing the word {string}' do |shouter, count, word|
  count.times do
    message = "A message containing the word #{word}"
    shout from: shouter, message: message
  end
end

Then "{person} should hear a shout" do |listener|
  expect(listener.messages_heard.count).to eq 1
end

Then "{person} should not hear a shout" do |person|
  expect(person.messages_heard.count).to eq 0
end

Then "{person} hears the following messages:" do |listener, expected_messages|
  actual_messages = listener.messages_heard.map { |message| [ message ] }

  expected_messages.diff!(actual_messages)
end

Then("{person} hears all {person}'s messages") do |listener, shouter|
  expect(listener.messages_heard).to match(messages_shouted_by[shouter.name])
end

Then("{person} should have {int} credits") do |shouter, credits|
  expect(shouter.credits).to eql(credits)
end
